autogpt_platform/backend/backend/data/notifications.py

```diff
@@ -0,0 +1,333 @@
+            NotificationType.MONTHLY_SUMMARY: BatchingStrategy.IMMEDIATE,
+        }
+        return BATCHING_RULES.get(self.notification_type, BatchingStrategy.HOURLY)
+
+    @property
+    def template(self) -> str:
+        """Returns template name for this notification type"""
+        return {
+            NotificationType.AGENT_RUN: "agent_run.html",
+            NotificationType.ZERO_BALANCE: "zero_balance.html",
+            NotificationType.LOW_BALANCE: "low_balance.html",
+            NotificationType.BLOCK_EXECUTION_FAILED: "block_failed.html",
+            NotificationType.CONTINUOUS_AGENT_ERROR: "agent_error.html",
+            NotificationType.DAILY_SUMMARY: "daily_summary.html",
+            NotificationType.WEEKLY_SUMMARY: "weekly_summary.html",
+            NotificationType.MONTHLY_SUMMARY: "monthly_summary.html",
+        }[self.notification_type]
+
+
+class NotificationPreference(BaseModel):
+    user_id: str
+    email: EmailStr
+    preferences: dict[NotificationType, bool] = {}  # Which notifications they want
+    daily_limit: int = 10  # Max emails per day
+    emails_sent_today: int = 0
+    last_reset_date: datetime = datetime.now()
+
+
+def get_batch_delay(notification_type: NotificationType) -> timedelta:
+    return {
+        NotificationType.AGENT_RUN: timedelta(seconds=1),
+        NotificationType.ZERO_BALANCE: timedelta(minutes=60),
+        NotificationType.LOW_BALANCE: timedelta(minutes=60),
+        NotificationType.BLOCK_EXECUTION_FAILED: timedelta(minutes=60),
+        NotificationType.CONTINUOUS_AGENT_ERROR: timedelta(minutes=60),
+    }[notification_type]
+
+
+async def create_or_add_to_user_notification_batch(
+    user_id: str,
+    notification_type: NotificationType,
+    data: str,  # type: 'NotificationEventModel'
+) -> dict:
+    logger.info(
+        f"Creating or adding to notification batch for {user_id} with type {notification_type} and data {data}"
+    )
+
+    notification_data = NotificationEventModel[
+        get_data_type(notification_type)
+    ].model_validate_json(data)
+
+    # Serialize the data
+    # serialized_data = json.dumps(notification_data.data.model_dump())
+    json_data: Json = Json(notification_data.data.model_dump_json())
+
+    # First try to find existing batch
+    existing_batch = await UserNotificationBatch.prisma().find_unique(
+        where={
+            "userId_type": {
+                "userId": user_id,
+                "type": notification_type,
+            }
+        },
+        include={"notifications": True},
+    )
+
+    if not existing_batch:
+        async with transaction() as tx:
+            notification_event = await tx.notificationevent.create(
+                data={
+                    "type": notification_type,
+                    "data": json_data,
+                }
+            )
+
+            # Create new batch
+            resp = await tx.usernotificationbatch.create(
```

Comment:
nit: you can do:
```suggestion
        async with transaction() as tx:
            notification_event = await NotificationEvent.prisma(tx).create(
                data={
                    "type": notification_type,
                    "data": json_data,
                }
            )

            # Create new batch
            resp = await NotificationBatch.prisma(tx).create(
```
which looks more like other DB code

Reply:
This pattern was explictly copied from other places transactions are used throughout the system lol